lib/services/identity.go

```diff
@@ -207,6 +207,21 @@ type Identity interface {
 	// GetUserTokenSecrets returns a user token secrets.
 	GetUserTokenSecrets(ctx context.Context, tokenID string) (types.UserTokenSecrets, error)
 
+	// UpsertRecoveryCodes upserts a user's new recovery codes.
+	UpsertRecoveryCodes(ctx context.Context, user string, recovery types.RecoveryCodesV1) error
+
+	// GetRecoveryCodes gets a user's recovery codes.
+	GetRecoveryCodes(ctx context.Context, user string) (*types.RecoveryCodesV1, error)
+
+	// CreateUserRecoveryAttempt logs user recovery attempt.
+	CreateUserRecoveryAttempt(ctx context.Context, user string, attempt types.RecoveryAttempt) error
+
+	// GetUserRecoveryAttempts returns user recovery attempts.
+	GetUserRecoveryAttempts(ctx context.Context, user string) ([]types.RecoveryAttempt, error)
```

Comment:
Suggestion: similarly to above, use a slice-to-pointer. Easier to reason about, IMO.

```suggestion
	GetUserRecoveryAttempts(ctx context.Context, user string) ([]*types.RecoveryAttempt, error)
```